rust/kernel/lib.rs

```diff
@@ -123,25 +123,44 @@ impl ThisModule {
     /// Locks the module parameters to access them.
     ///
     /// Returns a [`KParamGuard`] that will release the lock when dropped.
+    #[cfg(CONFIG_SYSFS)]
     pub fn kernel_param_lock(&self) -> KParamGuard<'_> {
         // SAFETY: `kernel_param_lock` will check if the pointer is null and
         // use the built-in mutex in that case.
-        #[cfg(CONFIG_SYSFS)]
-        unsafe {
-            bindings::kernel_param_lock(self.0)
-        }
+        unsafe { bindings::kernel_param_lock(self.0) }
 
         KParamGuard { this_module: self }
     }
+
+    /// Locks the module parameters to access them.
+    ///
+    /// Returns a [`KParamGuard`] that will release the lock when dropped.
+    #[cfg(not(CONFIG_SYSFS))]
+    pub fn kernel_param_lock(&self) -> KParamGuard {
+        // empty KParamGuard to keep the api uniform
+        KParamGuard { _private: () }
+    }
 }
 
 /// Scoped lock on the kernel parameters of [`ThisModule`].
 ///
 /// Lock will be released when this struct is dropped.
+#[cfg(CONFIG_SYSFS)]
+#[repr(transparent)]
 pub struct KParamGuard<'a> {
     this_module: &'a ThisModule,
 }
 
+/// Scoped lock on the kernel parameters of [`ThisModule`].
+///
+/// Lock will be released when this struct is dropped.
+#[cfg(not(CONFIG_SYSFS))]
+pub struct KParamGuard {
+    // private field, so this struct can not be constructed from outside this crate
+    #[cfg(not(CONFIG_SYSFS))]
+    _private: (),
+}
+
 #[cfg(CONFIG_SYSFS)]
 impl<'a> Drop for KParamGuard<'a> {
```

Comment:
Maybe add an empty drop impl for `not(CONFIG_SYSFS)`? The existence of a `Drop` impl affects the behavior of at least dropck by disallowing some code. It may be nice to keep the behavior the same with and without `CONFIG_SYSFS`.

Reply:
Can do that. Didn't know there was a different if there is a drop impl

Reply:
Now added an empty drop impl (it only touches the variable to make sure it cannot be optimised out).